tagsets/orsee_mysql.php

```diff
@@ -247,5 +247,121 @@ function dump_array($array,$title="",$dolang=true) {
+    }
+    if ($settings['database_version']<$system__database_version) {
+        $done=or_upgrade_database();
+        return $done;
+    } else {
+        return false;
+    }
+}
+
+function upgrade_database_version($new_version) {
+    global $settings;
+    $settings['database_version']=$new_version;
+    $done=orsee_db_save_array(array('option_value'=>$new_version),'options',1,'option_id');
+    if(!$done) log__admin("Database upgrade error. Could not set database version to new version number ".$new_version."!");
+    return $done;
+}
+
+
+function or_upgrade_database() {
+    global $settings, $system__database_version, $system__database_upgrades;
+    
+    if (!isset($system__database_upgrades)) {
+        $system__database_upgrades=array();
+    }
+
+    // check $system__database_upgrades for syntax errors, resort database upgrades by version number
+    $database_upgrades=array(); $i=0; $continue=true;
+    foreach ($system__database_upgrades as $upgr) {
+        $i++;
+        if ($continue) {
+            $error=or_upgrade_check_syntax($upgr);
+            if ($error) {
+                $continue=false;
+                log__admin("Error in upgrade syntax for $system__database_upgrades item nb ".$i.": ".$error." Upgrade aborted.");
+            } else {
+                $upgr['item_nb']=$i;
+                $database_upgrades[$upgr['version']][]=$upgr;
+            }
+        }
+    }
+    if ($continue) {
+        $done=ksort($database_upgrades,SORT_NUMERIC);
+    }
+    
+    // run the updates, stop if there is an error
+    $continue=true;
+    foreach ($database_upgrades as $this_version=>$vupgrades) {
+        if ($this_version>$settings['database_version']) {
+            foreach ($vupgrades as $upgr) {
+                if ($continue) {
+                   if ($upgr['type']=='new_lang_item') {
+                        $done=lang__upgrade_symbol_if_not_exists($upgr['specs']);
+                    } elseif ($upgr['type']=='new_admin_right') {
+                        $done=admin__update_admin_rights_if_not_exists($upgr['specs']);
+                    } elseif ($upgr['type']=='query') {
+                        $query=preg_replace('/TABLE\(([^)]+)\)/',table("$1"),$upgr['specs']['query_code']);
+                        $done=or_query($query);
+                        if ($done) {
+                            log__admin("Automatic database upgrade: executed query for \$system__database_upgrades item nb ".$i.".");
+                        } else {
+                            $continue=false;
+                            log__admin("Error in upgrade: Query for \$system__database_upgrades item nb ".$i." could not be executed. Upgrade aborted.");
+                        }
+                    }
+                }
+            }
+            if ($continue) {
+                $done=upgrade_database_version($this_version);
+                log__admin('Automatic database upgrade: Database upgraded to version '.$this_version.'.');
```

Comment:
If transaction is used, it needs to be committed when this $done is true. And it needs to be rolled-back if `$continue` is false.

Reply:
See above.